plugins/drivers/base/driver.go

```diff
@@ -0,0 +1,194 @@
+package base
+
+import (
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/hashicorp/nomad/client/allocdir"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/plugins/base"
+	"github.com/hashicorp/nomad/plugins/shared/hclspec"
+	"golang.org/x/net/context"
+)
+
+type DriverPlugin interface {
+	base.BasePlugin
+
+	TaskConfigSchema() (*hclspec.Spec, error)
+	Capabilities() (*Capabilities, error)
+	Fingerprint(context.Context) (<-chan *Fingerprint, error)
+
+	RecoverTask(*TaskHandle) error
+	StartTask(*TaskConfig) (*TaskHandle, error)
+	WaitTask(ctx context.Context, taskID string) (<-chan *ExitResult, error)
+	StopTask(taskID string, timeout time.Duration, signal string) error
+	DestroyTask(taskID string, force bool) error
+	InspectTask(taskID string) (*TaskStatus, error)
+	TaskStats(taskID string) (*TaskStats, error)
+	TaskEvents(context.Context) (<-chan *TaskEvent, error)
+
+	SignalTask(taskID string, signal string) error
+	ExecTask(taskID string, cmd []string, timeout time.Duration) (*ExecTaskResult, error)
+}
+
+type DriverSignalTaskNotSupported struct{}
+
+func (_ DriverSignalTaskNotSupported) SignalTask(taskID, signal string) error {
+	return fmt.Errorf("SignalTask is not supported by this driver")
+}
+
+type DriverExecTaskNotSupported struct{}
+
+func (_ DriverExecTaskNotSupported) ExecTask(taskID, signal string) error {
+	return fmt.Errorf("ExecTask is not supported by this driver")
+}
+
+type HealthState string
+
+var (
+	HealthStateUndetected = HealthState("undetected")
+	HealthStateUnhealthy  = HealthState("unhealthy")
+	HealthStateHealthy    = HealthState("healthy")
+)
+
+type Fingerprint struct {
+	Attributes        map[string]string
+	Health            HealthState
+	HealthDescription string
+}
+
+type FSIsolation string
+
+var (
+	FSIsolationNone   = FSIsolation("none")
+	FSIsolationChroot = FSIsolation("chroot")
+	FSIsolationImage  = FSIsolation("image")
+)
+
+type Capabilities struct {
+	// SendSignals marks the driver as being able to send signals
+	SendSignals bool
+
+	// Exec marks the driver as being able to execute arbitrary commands
+	// such as health checks. Used by the ScriptExecutor interface.
+	Exec bool
+
+	//FSIsolation indicates what kind of filesystem isolation the driver supports.
+	FSIsolation FSIsolation
+}
+
+type TaskConfig struct {
+	ID              string
+	Name            string
+	Env             map[string]string
+	Resources       Resources
+	Devices         []DeviceConfig
+	Mounts          []MountConfig
+	User            string
+	AllocDir        string
+	rawDriverConfig []byte
+}
+
+func (tc *TaskConfig) EnvList() []string {
+	l := make([]string, len(tc.Env))
+	for k, v := range tc.Env {
+		l = append(l, k+"="+v)
+	}
+	return l
```

Comment:
Use `sort.String(l)` to return env in a consistent order. Makes testing easier and improves UX.